src/settings.rs

```diff
@@ -0,0 +1,111 @@
+use std::{cell::RefCell, path::PathBuf};
+
+use anyhow::{anyhow, bail, Result};
+
+use crate::file;
+
+pub struct SettingsFile {
+    path: PathBuf,
+    cache: RefCell<Option<Settings>>,
+}
+
+impl SettingsFile {
+    pub fn new(path: PathBuf) -> Self {
+        Self {
+            path,
+            cache: RefCell::new(None),
+        }
+    }
+
+    fn write_settings(&self) -> Result<()> {
+        let s = self.cache.borrow().as_ref().unwrap().clone();
+        file::write_file(&self.path, &s.stringify())?;
+        Ok(())
+    }
+
+    fn read_settings(&self) -> Result<()> {
+        let mut needs_save = false;
+        {
+            let mut b = self.cache.borrow_mut();
+            if b.is_none() {
+                *b = Some(if file::is_file(&self.path) {
+                    let content = file::read_file("settings", &self.path)?;
+                    Settings::parse(&content)?
+                } else {
+                    needs_save = true;
+                    Default::default()
+                });
+            }
+        }
+        if needs_save {
+            self.write_settings()?;
+        }
+        Ok(())
+    }
+
+    pub fn with<T, F: FnOnce(&Settings) -> Result<T>>(&self, f: F) -> Result<T> {
+        self.read_settings()?;
+
+        // Settings can no longer be None so it's OK to unwrap
+        f(self.cache.borrow().as_ref().unwrap())
+    }
+
+    pub(crate) fn with_mut<T, F: FnOnce(&mut Settings) -> Result<T>>(&self, f: F) -> Result<T> {
+        self.read_settings()?;
+
+        // Settings can no longer be None so it's OK to unwrap
+        let result = { f(self.cache.borrow_mut().as_mut().unwrap())? };
+        self.write_settings()?;
+        Ok(result)
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct Settings {
+    pub default_toolchain: Option<String>,
+}
+
+fn get_value(table: &mut toml::value::Table, key: &str, path: &str) -> Result<toml::Value> {
+    table
+        .remove(key)
```

Comment:
Hmm this looks counter-intuitive given the name of the function - should the fn name be `take_value`? Similarly, `get_opt_string` might be better as `take_opt_string`?